Update WebOTP API docs (#27278)
mdn/content: files/en-us/web/api/webotp_api/index.md

{{securecontext_header}}{{DefaultAPISidebar("WebOTP API")}}

> The **WebOTP API** is an extension of the [Credential Management API](/en-US/docs/Web/API/Credential_Management_API) that enables web apps to verify that a phone number belongs to a user, for example when using it as a sign-in factor. This is done via a two-step process:

Author: Good point. I've updated it to 

> ... that provides a streamlined user experience for web apps to verify ...

1. The app client requests a one-time password (OTP), which is obtained from a specially-formatted SMS message sent by the app server.
2. This is then entered into a validation form on the app client and sent back to the server to verify that it matches what was originally sent in the SMS.

## WebOTP concepts and usage

Phone numbers are often used as a way to identify the user of an app. An SMS is frequently deployed to verify that the number belongs to the user. This typically contains an OTP that the user is required to copy and paste into a form on the app to verify that they own the number; a somewhat clunky user experience.

> Use cases include:

Author: "include" implies that there are others, and I feel that "but are not limited to" is a bit clumsy as language goes. I've updated it to 

> Example use cases include:

To round it out a bit and make it more obvious that these are just some examples, and not the whole list.

- Using a phone number as an account identifier instead of the traditional email address.
- Using a phone number as an extra factor (i.e. multifactor authentication) to improve sign-in security.
- Verification of sensitive actions, for example OK'ing a payment or transferring control of an account.

> The WebOTP API allows web apps to expedite this validation process by copying the OTP from the SMS and passing it to the app automatically after the user has provided consent. Most native platforms have an equivalent API.

Reviewer: Let's not forget to highlight the benefit of binding the OTP to the domain. It's not as strong constraint as WebAuthn, but can help the user figure the OTP is sent to the right domain.
cc: @yi-gu

Author: I've added a paragraph to cover this, inspired by some of the stuff in your article:

> Note that the OTP is also bound to the sending domain. This is a useful security constraint for verifying that the OTP is coming from the right source that may be enough for day-to-day reauthentication. You are, however, recommended to combine it with a stronger form of authentication such as the [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) for riskier actions like sign-up — phone numbers can be hijacked, or recycled by carriers.

Does that work?

Author: OK, no problem. I've moved the second part of the paragraph down to a note at the bottom of the "How does it work?" section. Does that work for you? I was a bit unsure about adding a whole new section just for a single caution, but I could be persuaded, if you think it needs to stand out a bit more.

Reviewer: I've suggested alternative wording, but it can be moved right below this quote given the section is within a "note".

### How does it work?

The process works like so:

1. At the point where phone number verification is required, an app client will ask a user to enter their phone number into a form, which is then submitted to the app server.

> 2. The app client then invokes {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} with an `otp` option, to request an OTP from the underlying system when a specially-formatted SMS message is received from the app server. This SMS should contain the app's domain to allow the system to identify the correct message, and the OTP code. The `get()` call is {{jsxref("Promise")}}-based, and waits for the SMS message to be sent.

Author: There is a lot going on in this paragraph! I made a few updates to Step 2, to streamline the text and mention the specified SMS transport type. Do you think this is better:

> The app client then invokes {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} with an `otp` option specifying a `transport` type of `"sms"`. This triggers a request for an OTP from the underlying system, the source of which will be a specially-formatted SMS message (containing the OTP and the app's domain) received from the app server. The `get()` call is {{jsxref("Promise")}}-based, and waits for the SMS message to be received.

?

> 2. The app client then invokes {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} with an `otp` option, to request an OTP from the underlying system when a specially-formatted SMS message is received from the app server. This SMS should contain the app's domain to allow the system to identify the correct message, and the OTP code. The `get()` call is {{jsxref("Promise")}}-based, and waits for the SMS message to be sent.

Reviewer 2: Sounds good!

3. Just after Step 2 has occurred, the app server should send the SMS message to the specified phone number.

> 4. When the SMS is received on the device, provided it contains the app's domain the browser will display a dialog asking the user if they consent to the OTP being retrieved. If they do consent, the `get()` call will fulfill with an {{domxref("OTPCredential")}} object containing the OTP.

Author: Ah, OK, so we want to state that this step is to ask for user consent, but some browsers may do it different to how Chrome does.

I've updated Step 4 like so:

> When the SMS is received on the device, provided it contains the app's domain the browser will ask the user if they consent to the OTP being retrieved/used. Chrome, for example, displays a dialog asking them for their permission to retrieve the OTP from the SMS; other browsers may handle it differently. If they do consent, the `get()` call will fulfill with an {{domxref("OTPCredential")}} object containing the OTP.

5. You can then use the OTP in whatever way you wish. Typical usage would be to set it as the value of the validation form on the app client and then submit the form, making the process as seamless as possible.
6. The app server will then verify that the OTP sent back to it matches what it originally sent in the SMS and, if so, complete the process (for example, sign the user in).

### The SMS message format

A typical SMS message looks like so:

```
Your verification code is 123456.

@www.example.com #123456
```

- The first and second lines are optional, and are for human-readability.

> - The last line is mandatory. It must be the line line if there are others present, and must consist of:

Author: Fixed; thanks!

You could specify a directive allowing use of it in a specific cross-origin domain (i.e.inside an {{htmlelement("iframe")}}) like this:

```http
Permissions-Policy: otp-credentials=(self "https://embedded.com")
```

Or could you specify it directly on the `<iframe>` like this:

```html
<iframe src="https://embedded.com/..." allow="otp-credentials"> ... </iframe>
```

> **Note:** Where a policy forbids use WebOTP `get()`, the {{jsxref("Promise", "promises")}} returned by it will reject with a `SecurityError` {{domxref("DOMException")}}.

## Interfaces

- {{domxref("OTPCredential")}}
  - : Returned when a WebOTP `get()` call fulfills; includes a `code` property that contains the retrieved OTP.

## Extensions to other interfaces

- {{domxref("CredentialsContainer.get()")}}, the `otp` option
  - : Calling `get()` with an `otp` option instructs the user agent to attempt to retrieve an OTP from the underlying system's SMS app.

## Examples

In this example, when an SMS message arrives, and the user grants permission, an {{domxref("OTPCredential")}} object is returned with an OTP. This password is then prefilled into the verification form field, and the form is submitted.

> [Try the demo using a phone here](https://glitch.com/edit/#!/web-otp?path=views%2Findex.html%3A55%3A8).

Reviewer: The demo URL should be https://web-otp.glitch.me/ .
https://glitch.com/edit/#!/web-otp?path=views%2Findex.html%3A55%3A8 points at the code editor.

Author: Good call. I've fixed this, both in the WebOTP page, and the `CredentialsContainer.get()` page.

3. The app server sends the SMS message to the specified phone number. This must be done just after Step 2 has occurred.
4. When the SMS is received on the device, provided it contains the app's domain the browser will ask the user if they consent to the OTP being retrieved/used. Chrome, for example, displays a dialog asking them for their permission to retrieve the OTP from the SMS; other browsers may handle it differently. If they do consent, the `get()` call will fulfill with an {{domxref("OTPCredential")}} object containing the OTP.
5. You can then use the OTP in any way you wish. Typical usage would be to set it as the value of the validation form on the app client and then submit the form, making the process as seamless as possible.
6. The app server will then verify that the OTP sent back to it matches what it originally sent in the SMS and, if so, complete the process (for example, sign the user in).

> > **Note:** You are recommended to combine WebOTP with a stronger form of authentication such as the [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) for actions like initial user sign-up — there is always the risk that phone numbers can be hijacked or recycled by carriers.

Reviewer: ```suggestion
> **Note:** Attackers can spoof SMS and hijack a person's phone number. Carriers can also recycle phone numbers to new users after an account is closed. While SMS OTP is useful to verify a phone number, you are recommended to use a stronger form of authentication such as the [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API).
```

{{securecontext_header}}{{DefaultAPISidebar("WebOTP API")}}

> The **WebOTP API** is an extension of the [Credential Management API](/en-US/docs/Web/API/Credential_Management_API) that provides a streamlined user experience for web apps to verify that a phone number belongs to a user, for example when using it as a sign-in factor. This is done via a two-step process:

Author: I wasn't quite sure about this edit, so I tweaked it a little, to:

> The **WebOTP API** provides a streamlined user experience for web apps to verify that a phone number belongs to a user when using it as a sign-in factor. WebOTP is an extension of the [Credential Management API](/en-US/docs/Web/API/Credential_Management_API).
> 
> The verification is done via a two-step process:

I think it is definitely better for being split up, but I've also put the sentence about it being an extension as the second sentence. Makes more sense to lead with the purpose of the API.

> The **WebOTP API** is an extension of the [Credential Management API](/en-US/docs/Web/API/Credential_Management_API) that provides a streamlined user experience for web apps to verify that a phone number belongs to a user, for example when using it as a sign-in factor. This is done via a two-step process:

Reviewer 3: Yes, this flow of thoughts is definitely much better. Thanks 👍

1. The app client requests a one-time password (OTP), which is obtained from a specially-formatted SMS message sent by the app server.

> 2. This is then entered into a validation form on the app client and sent back to the server to verify that it matches what was originally sent in the SMS.

Author: I've updated it to 

> 2. JavaScript is used to enter the OTP into a validation form on the app client and it is submitted back to the server to verify that it matches what was originally sent in the SMS.

## WebOTP concepts and usage

Phone numbers are often used as a way to identify the user of an app. An SMS is frequently deployed to verify that the number belongs to the user. This typically contains an OTP that the user is required to copy and paste into a form on the app to verify that they own the number; a somewhat clunky user experience.

> Example use cases include:

Author: This edit seemed a bit over the top. I've updated it to:

> OTP use cases include:

- Using a phone number as an extra factor (i.e. for two-factor authentication (2FA) or multifactor authentication (MFA)) to improve sign-in security.

> - Verification of sensitive actions, for example, verifying a payment.

Reviewer 3: The final version is definitely clearer 👍

The WebOTP API allows web apps to expedite this validation process by copying the OTP from the SMS and passing it to the app automatically after the user has provided consent (most native platforms have an equivalent API).

Note that the OTP is also bound to the sending domain. This is a useful security constraint for verifying that the OTP is coming from the right source, which can mitigate the risk of phishing attacks during day-to-day reauthentication.

### Security concerns with SMS OTPs

> SMS OTPs are useful for verifying phone numbers, and using SMS for a second factor is certainly better than having no second factor. In some regions, other identifiers such as email addresses and authenticators are not widely-used, so SMS OTPs are very common.

Author: No; in authentication it is common to talk about adding a second factor to improve security, thereby creating an auth experience that uses two-factor authentication (or multi-factor authentication)

However, SMS isn't that secure. Attackers can spoof SMS and hijack a person's phone number, and carriers can recycle phone numbers to new users after an account is closed.

You are therefore recommended to use a stronger form of authentication if possible, such as a [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API)-based solution involving a password and security key, or a passkey.

> ### How does it work?

Author: Better, definitely. I just added "the" before the API name, and made it into an H2. Because more recently I've been including these "How does xxx work?" sections on every landing page, and I think it is nice to be able to find them form the right hand nav.

5. You can then use the OTP in any way you wish. Typical usage would be to set it as the value of the validation form on the app client and then submit the form, making the process as seamless as possible.
6. The app server will then verify that the OTP sent back to it matches what it originally sent in the SMS and, if so, complete the process (for example, sign the user in).

### The SMS message format

A typical SMS message looks like so:

```
Your verification code is 123456.

@www.example.com #123456
```

- The first and second lines are optional, and are for human-readability.
- The last line is mandatory. It must be the last line if there are others present, and must consist of:
  - The domain part of the URL of the website that invoked the API, preceded by a `@`.
  - Followed by a space.

>   - Followed by the OTP, preceded by a pound sign (`#`).

Reviewer 3: I see only two lines of text here: "Your verification..." and "@www...". Is the second line blank here and should we state that?

Author: Sure. I've updated the first bullet to

> - The first line and second blank line are optional and are for human readability.

> **Note:** The provided domain value must not include a URL schema, port, or other URL features not shown above.

> If the `get()` method is being invoked by a 3rd party site embedded in an {{htmlelement("iframe")}}, the SMS should be structured as followed:

Reviewer 3: ```suggestion
If the `get()` method is invoked by a third-party site embedded in an {{htmlelement("iframe")}}, the SMS structure should be:
```

```
Your verification code is 123456.

@top-level.example.com #123456 @embedded.com
```

In this case, the last line must consist of:

- The domain part of the top-level domain, preceded by a `@`.
- Followed by a space.
- Followed by the OTP, preceded by a pound sign (`#`).
- Followed by a space.
- Followed by the domain part of the embedded domain, preceded by a `@`.

## Controlling access to the API

> The availability of WebOTP can be controlled using a [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) specifying a {{httpheader("Permissions-Policy/otp-credentials", "otp-credentials")}} directive. This directive has a default allowlist value of `"self"`, meaning that by default these methods can be used in top-level document contexts.
> 
> You could specify a directive allowing use of it in a specific cross-origin domain (i.e.inside an {{htmlelement("iframe")}}) like this:

Reviewer 3: can we replace "it"?

```suggestion
The availability of WebOTP can be controlled using a [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy) specifying a {{httpheader("Permissions-Policy/otp-credentials", "otp-credentials")}} directive. This directive has a default allowlist value of `"self"`, meaning that by default, these methods can be used in top-level document contexts.

You could specify a directive allowing the use of it in a specific cross-origin domain (i.e., inside an {{htmlelement("iframe")}}) like this:
```

Author: Yeah, I've put "WebOTP":

> You could specify a directive allowing the use of WebOTP in a specific cross-origin domain (i.e., inside an {{htmlelement("iframe")}}) like this:

```http
Permissions-Policy: otp-credentials=(self "https://embedded.com")
```

> Or could you specify it directly on the `<iframe>` like this:

Reviewer 3: ```suggestion
Or you could specify it directly on the `<iframe>` like this:
```

```html
<iframe src="https://embedded.com/..." allow="otp-credentials"> ... </iframe>
```

> > **Note:** Where a policy forbids use WebOTP `get()`, the {{jsxref("Promise", "promises")}} returned by it will reject with a `SecurityError` {{domxref("DOMException")}}.

Reviewer 3: - to use?
- can we replace "it"?

```suggestion
> **Note:** Where a policy forbids to use WebOTP `get()`, the {{jsxref("Promise", "promises")}} returned by it will reject with a `SecurityError` {{domxref("DOMException")}}.
```

## Interfaces

- {{domxref("OTPCredential")}}
  - : Returned when a WebOTP `get()` call fulfills; includes a `code` property that contains the retrieved OTP.

## Extensions to other interfaces

- {{domxref("CredentialsContainer.get()")}}, the `otp` option
  - : Calling `get()` with an `otp` option instructs the user agent to attempt to retrieve an OTP from the underlying system's SMS app.

## Examples

> In this example, when an SMS message arrives, and the user grants permission, an {{domxref("OTPCredential")}} object is returned with an OTP. This password is then prefilled into the verification form field, and the form is submitted.
> 
> [Try the demo using a phone here](https://web-otp.glitch.me/).

Reviewer 3: ```suggestion
In this example, when an SMS message arrives and the user grants permission, an {{domxref("OTPCredential")}} object is returned with an OTP. This password is then prefilled into the verification form field, and the form is submitted.

[Try this demo using a phone](https://web-otp.glitch.me/).
```

> The form field includes an [`autocomplete`](/en-US/docs/Web/HTML/Attributes/autocomplete) attribute with a value of `one-time-code`. This is not needed for the WebOTP API to work, but it is worth including. As a result, Safari will prompt the user to autofill this field with the OTP when a correctly-formatted SMS is received, even thought it doesn't fully support the WebOTP API.

Reviewer 3: ```suggestion
The form field includes an [`autocomplete`](/en-US/docs/Web/HTML/Attributes/autocomplete) attribute with the value of `one-time-code`. This is not needed for the WebOTP API to work, but it is worth including. As a result, Safari will prompt the user to autofill this field with the OTP when a correctly-formatted SMS is received, even though the WebOTP API isn't fully supported in Safari.
```

```html
<input type="text" autocomplete="one-time-code" inputmode="numeric" />
```

The JavaScript is as follows:

```js

> // Feature detect support via OTPCredential availability

Reviewer 3: should this be 
```suggestion
// Detect feature support via OTPCredential availability
```

}
```

Another good use for the {{domxref("AbortController")}} is to cancel the `get()` request after a certain amount of time:

```js
setTimeout(() => {
  // abort after 30 seconds
  ac.abort();
}, 30 * 1000);
```

If the user becomes distracted or navigates somewhere else, it is good to cancel the request so that they don't get presented with a permission prompt that is no longer relevant to them.

## Specifications

{{Specifications}}

## See also

> - [Verify phone numbers on the web with WebOTP](https://developer.chrome.com/articles/web-otp/)

Reviewer 3: ```suggestion
- [Verify phone numbers on the web with WebOTP](https://developer.chrome.com/articles/web-otp/) on developer.chrome.com (2023)
```
